network/src/gossip/message_routing.rs

```diff
@@ -0,0 +1,335 @@
+					)
+				}
+			}
+		}
+	}
+
+	/// Whether a message with given topic is live.
+	pub fn is_topic_live(&self, topic: &Hash) -> bool {
+		self.expected_queues.get(topic).is_some()
+	}
+
+	/// Whether a message is allowed under the intersection of the given leaf-set
+	/// and our own.
+	pub fn allowed_intersecting(&self, other_leaves: &LeavesVec, topic: &Hash) -> bool {
+		for i in other_leaves {
+			for j in &self.leaves {
+				if i == j {
+					let leaf_topics = self.leaf_topics.get(i)
+						.expect("leaf_topics are mutated only in update_leaves; \
+							we have an entry for each item in self.leaves; \
+							i is in self.leaves; qed");
+
+					if leaf_topics.contains(topic) {
+						return true;
+					}
+				}
+			}
+		}
+
+		false
+	}
+
+	/// Get topics of all message queues a peer is interested in - this is useful
+	/// when a peer has informed us of their new best leaves.
+	pub fn intersection_topics(&self, other_leaves: &LeavesVec) -> impl Iterator<Item=Hash> {
+		let deduplicated = other_leaves.iter()
+			.filter_map(|l| self.leaf_topics.get(l))
+			.flat_map(|topics| topics.iter().cloned())
+			.collect::<HashSet<_>>();
+
+		deduplicated.into_iter()
+	}
+
+	/// Iterate over all live message queues for which the data is marked as not locally known,
+	/// calling a closure with `(topic, root)`. The closure will return whether the queue data is
+	/// unknown.
+	///
+	/// This is called when we should send un-routed message queues that we are
+	/// newly aware of to peers - as in when we update our leaves.
+	pub fn sweep_unknown_queues(&mut self, mut check_known: impl FnMut(&Hash, &Hash) -> bool) {
+		for (topic, &mut (ref queue_root, ref mut known)) in self.expected_queues.iter_mut() {
+			if !*known {
+				*known = check_known(topic, queue_root)
+			}
+		}
+	}
+}
+
+#[cfg(test)]
+mod tests {
+	use super::*;
+	use crate::tests::TestChainContext;
+	use crate::gossip::{Known, GossipParachainMessages};
+	use polkadot_primitives::parachain::Message as ParachainMessage;
+
+	fn hash(x: u8) -> Hash {
+		[x; 32].into()
+	}
+
+	fn message_queue(from: u8, to: u8) -> Option<[[u8; 2]; 1]> {
+		if from == to {
+			None
+		} else {
+			Some([[from, to]])
+		}
+	}
+
+	fn message_queue_root(from: u8, to: u8) -> Option<Hash> {
+		message_queue(from, to).map(
+			|q| polkadot_validation::message_queue_root(q.iter())
+		)
+	}
+
+	fn check_roots(view: &mut View, i: u8, max: u8) -> bool {
+		for j in 0..max {
+			if let Some(messages) = message_queue(i, j) {
+				let queue_root = message_queue_root(i, j).unwrap();
+				let messages = GossipParachainMessages {
+					queue_root,
+					messages: messages.iter().map(|m| ParachainMessage(m.to_vec())).collect(),
+				};
+
+				match view.validate_queue_and_note_known(&messages).0 {
+					GossipValidationResult::ProcessAndKeep(topic) => if topic != queue_topic(queue_root) {
+						return false
+					},
+					_ => return false,
+				}
+			}
+		}
+
+		true
+	}
+
+	#[test]
+	fn update_leaves_none_in_common() {
```

Comment:
Generally, these tests may explain a lot about what's going on if their intent and logic are documented a bit more thoroughly.